addons/sale_purchase/models/sale_order_line.py

```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
-        PurchaseOrder = self.env['purchase.order']
         supplier_po_map = {}
-        sale_line_purchase_map = {}
+        purchase_lines = self.env['purchase.order.line']
         for line in self:
-            line = line.with_company(line.company_id)
-            # determine vendor of the order (take the first matching company and product)
-            suppliers = line.product_id._select_seller(quantity=line.product_uom_qty, uom_id=line.product_uom)
-            if not suppliers:
-                raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (line.product_id.display_name,))
-            supplierinfo = suppliers[0]
+            line = line.with_company(line._purchase_service_get_company())
+            supplierinfo = self._purchase_service_match_supplier()
             partner_supplier = supplierinfo.partner_id
 
             # determine (or create) PO
             purchase_order = supplier_po_map.get(partner_supplier.id)
             if not purchase_order:
-                purchase_order = PurchaseOrder.search([
-                    ('partner_id', '=', partner_supplier.id),
-                    ('state', '=', 'draft'),
-                    ('company_id', '=', line.company_id.id),
-                ], limit=1)
-            if not purchase_order:
-                values = line._purchase_service_prepare_order_values(supplierinfo)
-                purchase_order = PurchaseOrder.with_context(mail_create_nosubscribe=True).create(values)
-            else:  # update origin of existing PO
+                purchase_order = self._match_or_create_purchase_order(supplierinfo)
+            else:  # if not already updated origin in this loop
                 so_name = line.order_id.name
-                origins = []
-                if purchase_order.origin:
-                    origins = purchase_order.origin.split(', ') + origins
+                origins = (purchase_order.origin or '').split(', ')
                 if so_name not in origins:
-                    origins += [so_name]
-                    purchase_order.write({
-                        'origin': ', '.join(origins)
-                    })
-            supplier_po_map[partner_supplier.id] = purchase_order
+                    purchase_order.write({'origin': ', '.join(origins + [so_name])})
+                supplier_po_map[partner_supplier.id] = purchase_order
 
             # add a PO line to the PO
             values = line._purchase_service_prepare_line_values(purchase_order, quantity=quantity)
-            purchase_line = line.env['purchase.order.line'].create(values)
+            purchase_lines |= line.env['purchase.order.line'].create(values)
 
-            # link the generated purchase to the SO line
-            sale_line_purchase_map.setdefault(line, line.env['purchase.order.line'])
-            sale_line_purchase_map[line] |= purchase_line
-        return sale_line_purchase_map
+        return purchase_lines
```

Comment:
Here the signature of the method has changed, as we return a recordset instead of a dictionary SO line => PO line.
Maybe this result is not used anywhere in our code, but it could be used by third-party modules.

Reply:
reverted to minimal diff, but a lot of useless code...
As discuss, it was in un-deployed saas version, so the risk of override of this function is really weak